vim25/xml/util/util_examples_test.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright (c) 2021 VMware, Inc. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package util implements some helper functions for marshaling/unmarshaling
+// vim objects to/from XML.
+package util_test
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/govmomi/vim25/xml"
+	"github.com/vmware/govmomi/vim25/xml/util"
+)
+
+func ExampleVimObjectsFromString() {
+	const vimObjects = `
+<obj xmlns:vim25="urn:vim25"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="vim25:VirtualMachineConfigSpec">
+    <name>go-vm</name>
+    <numCPUs>2</numCPUs>
+    <memoryMB>2048</memoryMB>
+</obj>
+<obj xmlns:vim25="urn:vim25"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="vim25:VirtualMachineConfigInfo">
+    <name>go-vm</name>
+    <version>vmx-07</version>
+    <guestId>dosGuest</guestId>
+    <hardware>
+        <numCPU>2</numCPU>
+        <memoryMB>2048</memoryMB>
+    </hardware>
+</obj>
+`
+
+	// Decode all of the objects from the XML string.
+	chanObj, chanErr := util.VimObjectsFromString(vimObjects)
+
+	// For each object we see, print its type and information about it, along
+	// with any errors that occur along the way.
+	for {
+		select {
+		case obj := <-chanObj:
+			if obj == nil {
```

Comment:
IMHO this pattern can lead to a situation where the `chanErr` is not correctly drained. A closed channel always returns the zero value (causing hot loop and thus likelihood of being scheduled before the other channel has a chance to fire).

In `util.go` once the goroutine loop is done, `chanObj` is closed before `chanErr`. So this would return too early (same in `util_test.go`).

The general pattern in these cases is to instruct the `for` loop to stop on a condition variable and set the channel to `nil` once it's closed, e.g. pseudo-code:

```go
// You can edit this code!
// Click here and start typing.
package main

import "time"

func main() {
	var (
		errChan chan error
		objChan chan interface{}
	)

	go func() {
		time.Sleep(2 * time.Second)
		close(errChan)
		close(objChan)
	}()

	for objChan != nil && errChan != nil {
		select {
		case o, ok := <-objChan:
			if !ok {
				objChan = nil

			}

		case err, ok := <-errChan:
			if !ok {
				errChan = nil
			}
		}
	}
}
```

Since the receiver doesn't have control over the channel in your case (read-only), someone has to set it to `nil` or the channel must be made writeable.

Reply:
Thanks for the review. I knew when I wrote the code that I was forgetting something about channels (the implicit tuple they return with the `ok bool` if the channel is closed), but I just didn't spend too much time on it. As for the other pattern you recommended:

* I do not really care too much about the draining here since it is example code and I know exactly how it will behave
* I've never honestly seen this patter, but regardless, I do not think it is necessary here
* Even if a channel is closed, it can still be drained by a reader. See this Go playground [example](https://go.dev/play/p/oXo2tduEMv0)

    ```go
    package main
    
    import (
    	"fmt"
    	"time"
    )
    
    func main() {
    	c := make(chan int)
    	go func() {
    		c <- 1
    		close(c)
    	}()
    	time.Sleep(3 * time.Second)
    	a, ok := <-c
    	fmt.Printf("%d %v\n", a, ok)
    }
    ```

    The above example will print:

    ```bash
    1 true
    ```

    Thus it is fine to close a channel if there is still data on that channel. It will be read until there is no more data.

Reply:
Just looked at the code again and since you're using an unbuffered `err` channel one issue I had in mind cannot occur here. My concern was not draining a closed channel but missing an error due to runtime scheduling. Unbuffered channel and your current logic makes this impossible IMHO 👍

Reply:
The above [example](https://go.dev/play/p/GhNppD6ZtWL) still works with buffered channels as well 😃 

```golang
package main

import (
	"fmt"
	"time"
)

func main() {
	c := make(chan int, 2)
	go func() {
		c <- 1
		c <- 2
		close(c)
	}()
	time.Sleep(3 * time.Second)
	a, ok := <-c
	fmt.Printf("%d %v\n", a, ok)
	a, ok = <-c
	fmt.Printf("%d %v\n", a, ok)
}
```

The above example prints:

```bash
1 true
2 true
```

Reply:
The point is that closing a channel does not prevent the channel from being drained.

Reply:
Sorry for not being clear. With buffered channels there's this issue https://go.dev/play/p/CqFQGdiqMlv.

Run it multiple times and you'll see. I just realized that this won't happen in your case since you're using unbuffered.

Reply:
I see what you are saying. This only matters if you actually want to ensure all of the information is read off all of the channels though. Either way, I get your point, but I still think it depends on the context of the situation, not whether or not the channel is buffered.